azure-iot-device/azure/iot/device/common/transport/mqtt/pipeline_stages.py

```diff
@@ -0,0 +1,645 @@
+            topic = event.topic
+            message = Message(event.payload)
+
+            if mqtt_topic_helper.is_c2d_topic(topic):
+                mqtt_topic_helper.extract_properties_from_topic(topic, message)
+                self.handle_pipeline_event(pipeline_events.IotHub_C2dMessage(message))
+
+            elif mqtt_topic_helper.is_input_topic(topic):
+                mqtt_topic_helper.extract_properties_from_topic(topic, message)
+                input_name = mqtt_topic_helper.get_input_name_from_topic(topic)
+                self.handle_pipeline_event(pipeline_events.IotHub_InputMessage(input_name, message))
+
+            else:
+                logger.warning("Warning: dropping message with topic %s", topic)
+
+        else:
+            # all other messages get passed up
+            PipelineStage._handle_pipeline_event(self, event)
+
+
+class Mqtt_Provider(PipelineStage):
+    """
+    PipelineStage object which is responsible for interfacing with the Mqtt provider object.
+    This stage handles all Mqtt_ operations and any other operations (such as Iot_Connect) that
+    can only be run at the protocol level.
+    """
+
+    def _run_op(self, op):
+        logger.info("{}({}) handling".format(self.name, op.name))
+
+        if op.name == pipeline_ops.names.Mqtt_SetConnectionArgs:
+            # Mqtt_SetConnectionArgs is where we create our MQTTProvider object and set
+            # all of its properties.
+            logger.info("{}({}) got connection args".format(self.name, op.name))
+            self.hostname = op.hostname
+            self.username = op.username
+            self.client_id = op.client_id
+            self.device_id = op.device_id
+            self.module_id = op.module_id
+            self.ca_cert = op.ca_cert
+            self.provider = MQTTProvider(
+                client_id=self.client_id,
+                hostname=self.hostname,
+                username=self.username,
+                ca_cert=self.ca_cert,
+            )
+            self.provider.on_mqtt_connected = self.on_connected
+            self.provider.on_mqtt_disconnected = self.on_disconnected
+            self.provider.on_mqtt_message_received = self._on_message_received
+            self.pipeline_root.provider = self.provider
+            self.complete_op(op)
+
+        elif op.name == pipeline_ops.names.Iot_SetSasToken:
+            # When we get a sas token from above, we just save it for later
+            logger.info("{}({}) got password".format(self.name, op.name))
+            self.sas_token = op.sas_token
+            self.complete_op(op)
+
+        elif op.name == pipeline_ops.names.Iot_Connect:
+            logger.info("{}({}) conneting".format(self.name, op.name))
+
+            def connect_callback():
+                self.provider.on_mqtt_connected = self.on_connected
+                self.on_connected()
+                self.complete_op(op)
+
+            self.provider.on_mqtt_connected = connect_callback
```

Comment:
>on_mqtt_connected [](start = 26, length = 17)

You reset the provider handler every time you call connect? Why?

Reply:
We always want self.on_connected to be called (it informs other stages that we're connected), but, if we're handling an Iot_Connect op, we also want to complete the op.  The easiest way to do that is to change on_mqtt_connected while we're waiting for the connection to complete.  Maybe there's a better way, I'll have to think about this.

---
In reply to: [279684136](https://github.com/Azure/azure-iot-sdk-python-preview/pull/87#discussion_r279684136) [](ancestors = 279684136)